### PR TITLE
not throw exception when add new items into cache folder

### PR DESCRIPTION
**Pull request by ansyral**
1. not throw exception when add new items into cache;
2. delete cache subfolder of current build if it failed

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs [hunk 1]
```diff
@@ -96,106 +96,123 @@ public void Build(IList<DocumentBuildParameters> parameters, string outputDirect
...
                         }
                     }
                 }
             }
+            catch (Exception)
+            {
+                if (_intermediateFolder != null)
+                {
+                    Directory.Delete(Path.Combine(Environment.ExpandEnvironmentVariables(_intermediateFolder), _currentBuildInfo.DirectoryName), true);
+                }
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: move to finally?

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs [hunk 2]
```diff
@@ -96,106 +96,123 @@ public void Build(IList<DocumentBuildParameters> parameters, string outputDirect
...
                         {
-                            Directory.Delete(Path.Combine(Environment.ExpandEnvironmentVariables(_intermediateFolder), _lastBuildInfo.DirectoryName), true);
+                            _currentBuildInfo.Save(_intermediateFolder);
+                            if (_lastBuildInfo != null)
+                            {
+                                try
+                                {
+                                    Directory.Delete(Path.Combine(Environment.ExpandEnvironmentVariables(_intermediateFolder), _lastBuildInfo.DirectoryName), true);
+                                }
+                                catch (Exception)
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: >Exception [](start = 39, length = 9)

IOException?

Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: >Exception [](start = 39, length = 9)

when will throw Exception? need we check existence here then delete?

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/BuildVersionInfo.cs [hunk 3]
```diff
@@ -139,5 +139,13 @@ internal void Save(string baseDir)
                 processor.Save(baseDir);
             }
         }
+
+        private void LoadWhenNotNull(string baseDir, string file, Action<string> action)
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/BuildVersionInfo.cs]: >LoadWhenNotNull [](start = 21, length = 15)

ActionWhenNotNull?

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs (hunk 2 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: IOException

---
In reply to: [106574086](https://github.com/dotnet/docfx/pull/1429#discussion_r106574086) [](ancestors = 106574086)

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs [hunk 4]
```diff
@@ -96,106 +96,123 @@ public void Build(IList<DocumentBuildParameters> parameters, string outputDirect
...
+                            }
                         }
                     }
                 }
             }
+            catch (Exception)
+            {
+                if (_intermediateFolder != null)
+                {
+                    Directory.Delete(Path.Combine(Environment.ExpandEnvironmentVariables(_intermediateFolder), _currentBuildInfo.DirectoryName), true);
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: catch IOException

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs (hunk 1 above)
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: no, only exception case need we delete current folder

---
In reply to: [106571262](https://github.com/dotnet/docfx/pull/1429#discussion_r106571262) [](ancestors = 106571262)

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs (hunk 2 above)
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: my idea is whatever exception is hit, no matter it is io, or pathtoolong, it shouldn't make the build fail

---
In reply to: [106574086](https://github.com/dotnet/docfx/pull/1429#discussion_r106574086) [](ancestors = 106574086)

Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: PathTooLongException is also IOException?

---
In reply to: [106577010](https://github.com/dotnet/docfx/pull/1429#discussion_r106577010) [](ancestors = 106577010,106574086)

**Comment by hellosnow** [comment]
:shipit:

**Comment by vwxyzh** [comment]
:shipit:

**Pull request merged**